PresentMonTraceConsumer.hpp

```diff
@@ -0,0 +1,359 @@
+
+    // Maps from queue packet submit sequence
+    // Used for Flip -> MMIOFlip -> VSyncDPC for FS, for PresentHistoryToken -> MMIOFlip -> VSyncDPC for iFlip,
+    // and for Blit Submission -> Blit completion for FS Blit
+    std::map<uint32_t, std::shared_ptr<PresentEvent>> mPresentsBySubmitSequence;
+
+    // Win32K present history tokens are uniquely identified by (composition surface pointer, present count, bind id)
+    // Using a tuple instead of named struct simply to have auto-generated comparison operators
+    // These tokens are used for "flip model" presents (windowed flip, dFlip, iFlip) only
+    typedef std::tuple<uint64_t, uint64_t, uint32_t> Win32KPresentHistoryTokenKey;
+    std::map<Win32KPresentHistoryTokenKey, std::shared_ptr<PresentEvent>> mWin32KPresentHistoryTokens;
+
+    // DxgKrnl present history tokens are uniquely identified by a single pointer
+    // These are used for all types of windowed presents to track a "ready" time
+    std::map<uint64_t, std::shared_ptr<PresentEvent>> mDxgKrnlPresentHistoryTokens;
+
+    // Present by window, used for determining superceding presents
+    // For windowed blit presents, when DWM issues a present event, we choose the most recent event as the one that will make it to screen
+    std::map<uint64_t, std::shared_ptr<PresentEvent>> mPresentByWindow;
+
+    // Presents that will be completed by DWM's next present
+    std::deque<std::shared_ptr<PresentEvent>> mPresentsWaitingForDWM;
+    // Used to understand that a flip event is coming from the DWM
+    uint32_t DwmPresentThreadId = 0;
+
+    // Windows that will be composed the next time DWM presents
+    // Generated by DWM events indicating it's received tokens targeting a given hWnd
+    std::set<uint32_t> mWindowsBeingComposed;
+
+    // Yet another unique way of tracking present history tokens, this time from DxgKrnl -> DWM, only for legacy blit
+    std::map<uint64_t, std::shared_ptr<PresentEvent>> mPresentsByLegacyBlitToken;
+
+    std::mutex mProcessMutex;
+    std::map<uint32_t, ProcessInfo> mNewProcessesFromETW;
+    std::vector<uint32_t> mDeadProcessIds;
+
+    bool DequeuePresents(std::vector<std::shared_ptr<PresentEvent>>& outPresents)
+    {
+        if (mCompletedPresents.size())
+        {
+            auto lock = scoped_lock(mMutex);
+            outPresents.swap(mCompletedPresents);
+            return !outPresents.empty();
+        }
+        return false;
+    }
+
+    void GetProcessEvents(decltype(mNewProcessesFromETW)& outNewProcesses, decltype(mDeadProcessIds)& outDeadProcesses)
+    {
+        auto lock = scoped_lock(mProcessMutex);
+        outNewProcesses.swap(mNewProcessesFromETW);
+        outDeadProcesses.swap(mDeadProcessIds);
+    }
+
+    virtual void OnEventRecord(_In_ PEVENT_RECORD pEventRecord);
+    virtual bool ContinueProcessing() { return !g_Quit; }
+
+private:
+    void CompletePresent(std::shared_ptr<PresentEvent> p)
```

Comment:
I'd put the private methods in the cpp. They were originally tiny helpers, and grew kind of organically into relatively large methods.